pkg/scheduler/objects/preemption.go

```diff
@@ -0,0 +1,875 @@
+				if queueSnapshot.IsAtOrAboveGuaranteedResource() {
+					// removing task does not violate queue constraints, adjust queue and node
+					nodeCurrentAvailable.AddTo(victim.GetAllocatedResource())
+					// check if ask now fits and we haven't had this happen before
+					if nodeCurrentAvailable.FitInMaxUndef(p.ask.GetAllocatedResource()) && index < 0 {
+						index = len(results)
+					}
+					// add victim to results
+					results = append(results, victim)
+				} else {
+					// add back resources
+					queueSnapshot.AddAllocation(victim.GetAllocatedResource())
+				}
+			}
+		}
+	}
+
+	// check to see if enough resources were freed
+	if index < 0 {
+		return -1, nil
+	}
+
+	return index, results
+}
+
+func (p *Preemptor) duplicateQueueSnapshots() map[string]*QueuePreemptionSnapshot {
+	cache := make(map[string]*QueuePreemptionSnapshot, 0)
+	for _, snapshot := range p.allocationsByQueue {
+		snapshot.Duplicate(cache)
+	}
+	return cache
+}
+
+// checkPreemptionPredicates calls the shim via the SI to evaluate nodes for preemption
+func (p *Preemptor) checkPreemptionPredicates(predicateChecks []*si.PreemptionPredicatesArgs, victimsByNode map[string][]*Allocation) *predicateCheckResult {
+	// don't process empty list
+	if len(predicateChecks) == 0 {
+		return nil
+	}
+
+	// sort predicate checks by number of expected preempted tasks
+	sort.SliceStable(predicateChecks, func(i int, j int) bool {
+		return predicateChecks[i].StartIndex < predicateChecks[j].StartIndex
+	})
+
+	// check for RM callback
+	plugin := plugins.GetResourceManagerCallbackPlugin()
+	if plugin == nil {
+		// if a plugin isn't registered, assume checks will succeed and synthesize a result
+		check := predicateChecks[0]
+		log.Logger().Debug("No RM callback plugin registered, using first selected node for preemption",
+			zap.String("NodeID", check.NodeID),
+			zap.String("AllocationKey", check.AllocationKey))
+
+		result := &predicateCheckResult{
+			allocationKey: check.AllocationKey,
+			nodeID:        check.NodeID,
+			success:       true,
+			index:         int(check.StartIndex),
+		}
+		result.populateVictims(victimsByNode)
+		return result
+	}
+
+	// process each batch of checks by sending to the RM
+	batches := batchPreemptionChecks(predicateChecks, preemptCheckConcurrency)
+	var bestResult *predicateCheckResult = nil
+	for _, batch := range batches {
+		var wg sync.WaitGroup
+		ch := make(chan *predicateCheckResult, len(batch))
+		expected := 0
+		for _, args := range batch {
+			// add goroutine for checking preemption
+			wg.Add(1)
+			expected++
+			go preemptPredicateCheck(plugin, ch, &wg, args)
+		}
+		// wait for completion and close channel
+		go func() {
+			wg.Wait()
+			close(ch)
+		}()
+		for result := range ch {
+			// if result is successful, keep track of it
+			if result.success {
+				if bestResult == nil {
+					bestResult = result
+				} else if result.betterThan(bestResult, p.allocationsByNode) {
+					bestResult = result
```

Comment:
codecov says this branch is not covered, is this relevant to cover?

Reply:
Added additional test in latest commit.